passes/cmds/splitfanout.cc

```diff
@@ -0,0 +1,244 @@
+					SigBit bit(sigmap(conn.second[i]));
+					bit_drivers_db[bit] = tuple<IdString,IdString,int>(cell->name, conn.first, i);
+				}
+			}
+		}
+
+		// Build bit_users_db and add edges to topological sorter
+		log("Building bit_users_db and adding edges to toposort...\n");
+		for (auto cell : module->cells()) {
+			for (auto conn : cell->connections()) {
+				if (!cell->input(conn.first)) continue;
+				for (int i = 0; i < GetSize(conn.second); i++) {
+					SigBit bit(sigmap(conn.second[i]));
+					if (!bit_drivers_db.count(bit)) continue;
+					bit_users_db[bit].insert(tuple<IdString,IdString,int>(cell->name,
+							conn.first, i-std::get<2>(bit_drivers_db[bit])));
+					IdString driver_cell = std::get<0>(bit_drivers_db[bit]);
+					if (toposort.has_node(driver_cell) && toposort.has_node(cell->name))
+						// toposort.edge(driver_cell, cell->name);
+						toposort.edge(cell->name, driver_cell);
+				}
+			}
+		}
+
+		// Build bit_users_db for output ports
+		log("Building bit_users_db for output ports...\n");
+		for (auto wire : module->wires()) {
+			if (!wire->port_output) continue;
+			SigSpec sig(sigmap(wire));
+			for (int i = 0; i < GetSize(sig); i++) {
+				SigBit bit(sig[i]);
+				if (!bit_drivers_db.count(bit)) continue;
+				bit_users_db[bit].insert(tuple<IdString,IdString,int>(wire->name,
+						IdString(), i-std::get<2>(bit_drivers_db[bit])));
+			}
+		}
+
+		// Sort using the topological sorter
+		log("Sorting using toposort...\n");
+		toposort.analyze_loops = false;
+		toposort.sort();
+	}
+
+	int split(Cell *cell, int limit)
+	{
+		// Get output signal/port
+		SigSpec outsig;
+		IdString outport;
+		int output_count = 0;
+		for (auto conn : cell->connections())
+			if (cell->output(conn.first)) {
+				output_count++;
+				outport = conn.first;
+				outsig = conn.second;
+			}
+		if (output_count != 1) {
+			log("Skipping %s cell %s/%s with %d output ports.\n", log_id(cell->type), log_id(module), log_id(cell), output_count);
+			return 0;
+		}
+		
+		// Check if output signal is "bit-split", skip if so
+		auto bit_users = bit_users_db[outsig[0]];
+		for (int i = 0; i < GetSize(outsig); i++) {
+			if (bit_users_db[outsig[i]] != bit_users) {
+				log("Skipping %s cell %s/%s with bit-split output.\n", log_id(cell->type), log_id(module), log_id(cell));
+				return 0;
+			}
+		}
+
+		// Skip if output signal has only one user
+		if (GetSize(bit_users) <= 1)
+			return 0;
+
+		// Skip if fanout is above limit
+		if (limit != -1 && GetSize(bit_users) > limit) {
+			log("Skipping %s cell %s/%s with high fanout %d.\n", log_id(cell->type), log_id(module), log_id(cell), GetSize(bit_users)-1);
+			return 0;
+		}
+
+		// Iterate over bit users and create a new cell for each one
+		log("Splitting %s cell %s/%s into %d copies based on fanout\n", log_id(cell->type), log_id(module), log_id(cell), GetSize(bit_users)-1);
+		int foi = 0;
+		cell->unsetPort(outport);
+		int num_new_cells = GetSize(bit_users)-1;
+		int bit_user_i = num_new_cells;
```

Comment:
What's the motivation for counting down here?

Reply:
The original reason was that in our fork, we used `bit_user_i` as a suffix for the new cell's name. But that doesn't apply here, as we just use `NEW_ID` for the new cell's name. It could be changed to increment if that is preferred